pkg/clients/frontegg_client.go

```diff
@@ -0,0 +1,238 @@
+		return "", "", time.Time{}, err
+	}
+
+	tokenString, ok := result["accessToken"].(string)
+	if !ok {
+		return "", "", time.Time{}, errors.New("access token not found in the response")
+	}
+
+	// Parse the token without verifying the signature.
+	token, _, err := new(jwt.Parser).ParseUnverified(tokenString, jwt.MapClaims{})
+	if err != nil {
+		return "", "", time.Time{}, fmt.Errorf("error parsing token: %v", err)
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		return "", "", time.Time{}, errors.New("invalid token claims")
+	}
+
+	email, ok := claims["email"].(string)
+	if !ok {
+		return "", "", time.Time{}, errors.New("email claim not found in token")
+	}
+
+	var tokenExpiry time.Time
+	if expiresIn, ok := result["expiresIn"].(float64); ok {
+		tokenExpiry = time.Now().Add(time.Duration(expiresIn) * time.Second)
+	} else {
+		// Default expiry time if not provided in the response
+		tokenExpiry = time.Now().Add(1 * time.Hour)
+	}
+
+	return tokenString, email, tokenExpiry, nil
+}
+
+// Get the token from the FronteggClient
+func (c *FronteggClient) GetToken() (string, error) {
+	return c.Token, nil
+}
+
+// Get the email from the FronteggClient
+func (c *FronteggClient) GetEmail() (string, error) {
+	return c.Email, nil
+}
+
+// Get the endpoint from the FronteggClient
+func (c *FronteggClient) GetEndpoint() (string, error) {
+	return c.Endpoint, nil
+}
+
+// Get the token expiry from the FronteggClient
+func (c *FronteggClient) GetTokenExpiry() (time.Time, error) {
+	return c.TokenExpiry, nil
+}
+
+// Get the password from the FronteggClient
+func (c *FronteggClient) GetPassword() (string, error) {
+	return c.Password, nil
+}
+
+func cloneRequest(r *http.Request) *http.Request {
+	// Deep copy the request
+	r2 := new(http.Request)
+	*r2 = *r
+	// Deep copy the URL
+	r2.URL = new(url.URL)
+	*r2.URL = *r.URL
+	// Deep copy the Header
+	r2.Header = make(http.Header, len(r.Header))
+	for k, s := range r.Header {
+		r2.Header[k] = append([]string(nil), s...)
+	}
+	return r2
+}
+
+func parseAppPassword(password string) (string, string, error) {
+	strippedPassword := strings.TrimPrefix(password, "mzp_")
+	var clientId, secretKey string
+
+	re := regexp.MustCompile("[^0-9a-fA-F]")
+	filteredChars := re.ReplaceAllString(strippedPassword, "")
+
+	if len(filteredChars) < 64 {
+		return "", "", fmt.Errorf("invalid app password length: %d", len(filteredChars))
+	}
+
+	clientId = formatDashlessUuid(filteredChars[0:32])
+	secretKey = formatDashlessUuid(filteredChars[32:])
+
+	return clientId, secretKey, nil
+}
+
+func formatDashlessUuid(dashlessUuid string) string {
+	parts := []string{
+		dashlessUuid[0:8],
+		dashlessUuid[8:12],
+		dashlessUuid[12:16],
+		dashlessUuid[16:20],
+		dashlessUuid[20:],
+	}
+	return strings.Join(parts, "-")
+}
+
+func (c *FronteggClient) NeedsTokenRefresh() bool {
+	return time.Now().After(c.TokenExpiry)
+}
+
+func (c *FronteggClient) RefreshToken() error {
+	log.Printf("[DEBUG] Refreshing Frontegg: %v\n", c)
```

Comment:
Did you intend to keep this?

Reply:
(This is just how you emit debug logs in Terraform Providers: https://developer.hashicorp.com/terraform/plugin/log/writing#debug-1)